### PR TITLE
txsource: delegation workload

### PR DESCRIPTION
**Pull request by ptrus**
Part of: #2506

Delegation workload continuously makes escrow and reclaim escrow transactions between accounts. (note: the escrow balance is only funded at start which effectively means reclaim is also tested as a part of this workload).

Also fixes a bug in delegation queries (extracted into #2756)

TODO:
- [x] probably should add unit tests for the delegation queries

### CODE REVIEW

File: go/oasis-node/cmd/debug/txsource/workload/delegation.go [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+		// Fund the account with delegation amount.
+		// Funds for fee's will be transferred before making transactions.
+		if err = transferFunds(ctx, delegationLogger, cnsc, fundingAccount, accounts[i].signer.Public(), delegateAmount); err != nil {
+			return fmt.Errorf("account funding failure: %w", err)
+		}
+	}
+
+	client := staking.NewStakingClient(conn)
+
+MAIN:
```
Comment (Yawning) [inline go/oasis-node/cmd/debug/txsource/workload/delegation.go]: Do you need this label?

File: go/oasis-node/cmd/debug/txsource/workload/delegation.go [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+
+		// Get current epoch.
+		epoch, err := cnsc.GetEpoch(ctx, consensus.HeightLatest)
+		if err != nil {
+			return fmt.Errorf("GetEpoch: %w", err)
+		}
+
+		var signedTx *transaction.SignedTransaction
+		var gas transaction.Gas
+		switch rng.Intn(2) {
```
Comment (Yawning) [inline go/oasis-node/cmd/debug/txsource/workload/delegation.go]: for maintainability, consider refactoring each case into a helper function

File: go/oasis-node/cmd/debug/txsource/workload/delegation.go (hunk 1 above)
Comment (ptrus) [reply go/oasis-node/cmd/debug/txsource/workload/delegation.go]: not anyome :+1:

File: go/oasis-node/cmd/debug/txsource/workload/delegation.go (hunk 2 above)
Comment (ptrus) [reply go/oasis-node/cmd/debug/txsource/workload/delegation.go]: done

File: go/consensus/tendermint/apps/staking/state/state_test.go [hunk 3]
```diff
@@ -14,9 +14,14 @@ import (
 	memorySigner "github.com/oasislabs/oasis-core/go/common/crypto/signature/signers/memory"
 	"github.com/oasislabs/oasis-core/go/common/quantity"
 	"github.com/oasislabs/oasis-core/go/consensus/tendermint/abci"
+	epochtime "github.com/oasislabs/oasis-core/go/epochtime/api"
 	staking "github.com/oasislabs/oasis-core/go/staking/api"
 )
 
+const (
```
Comment (Yawning) [inline go/consensus/tendermint/apps/staking/state/state_test.go]: Minor nitpick, can just have this as one line, and the constant definition should probably be in `TestDelegationQueries`

File: go/consensus/tendermint/apps/staking/state/state_test.go [hunk 4]
```diff
@@ -27,6 +32,85 @@ func mustInitQuantityP(t *testing.T, i int64) *quantity.Quantity {
...
+	// Generate escrow account.
+	escrowSigner, err := fac.Generate(signature.SignerEntity, rand.Reader)
+	require.NoError(t, err, "generating escrow signer")
+	escrowID := escrowSigner.Public()
+
+	escrowAccount := &staking.Account{}
+	s.SetAccount(escrowID, escrowAccount)
+
+	// Generate delegator accounts.
+	delegatorIDs := []signature.PublicKey{}
```
Comment (Yawning) [inline go/consensus/tendermint/apps/staking/state/state_test.go]: `var delegatorIDs []signature.PublicKey`

Similar comments from here omitted for brevity.

File: go/consensus/tendermint/apps/staking/state/state_test.go [hunk 5]
```diff
@@ -27,6 +32,85 @@ func mustInitQuantityP(t *testing.T, i int64) *quantity.Quantity {
...
+	s := NewMutableState(tree)
+
+	fac := memorySigner.NewFactory()
+
+	// Generate escrow account.
+	escrowSigner, err := fac.Generate(signature.SignerEntity, rand.Reader)
+	require.NoError(t, err, "generating escrow signer")
+	escrowID := escrowSigner.Public()
+
+	escrowAccount := &staking.Account{}
```
Comment (Yawning) [inline go/consensus/tendermint/apps/staking/state/state_test.go]: `var escrowAccount staking.Account`?

File: go/consensus/tendermint/apps/staking/state/state_test.go [hunk 6]
```diff
@@ -27,6 +32,85 @@ func mustInitQuantityP(t *testing.T, i int64) *quantity.Quantity {
...
+		signer, serr := fac.Generate(signature.SignerEntity, rand.Reader)
+		require.NoError(t, serr, "memory signer factory Generate account")
+		id := signer.Public()
+
+		delegatorIDs = append(delegatorIDs, id)
+
+		// Init account.
+		account := &staking.Account{}
+		account.General.Nonce = uint64(i)
+		require.NoError(t, account.General.Balance.FromBigInt(big.NewInt(2*i*100)), "initialize delegator account general balance")
```
Comment (Yawning) [inline go/consensus/tendermint/apps/staking/state/state_test.go]: As a matter of taste, and (really old habit), I dislike assertions with side-effects.  Rewrite to separate setting the balance, and the assertion.

(Similar comments omitted for brevity.)

File: go/consensus/tendermint/apps/staking/state/state_test.go [hunk 7]
```diff
@@ -27,6 +32,85 @@ func mustInitQuantityP(t *testing.T, i int64) *quantity.Quantity {
...
+func TestDelegationQueries(t *testing.T) {
+	db := dbm.NewMemDB()
+	tree := iavl.NewMutableTree(db, 128)
+	s := NewMutableState(tree)
+
+	fac := memorySigner.NewFactory()
+
+	// Generate escrow account.
+	escrowSigner, err := fac.Generate(signature.SignerEntity, rand.Reader)
+	require.NoError(t, err, "generating escrow signer")
```
Comment (Yawning) [inline go/consensus/tendermint/apps/staking/state/state_test.go]: Can save yourself a lot of typing by doing:
```
  require := require.New(t)

  require.NoError(err, "look ma! I saved having to specify `t` everywhere")
```

File: go/oasis-node/cmd/debug/txsource/workload/delegation.go [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+			break
+		}
+	}
+	if fromPermIdx == -1 {
+		d.logger.Debug("all accounts already delegating or debonding, skipping delegation")
+		return nil
+	}
+
+	// Select an account to delegate to.
+	// XXX: we could also cover the self-escrow case.
```
Comment (Yawning) [inline go/oasis-node/cmd/debug/txsource/workload/delegation.go]: So, do so?

File: go/consensus/tendermint/apps/staking/state/state_test.go (hunk 3 above)
Comment (ptrus) [reply go/consensus/tendermint/apps/staking/state/state_test.go]: yeah, done

File: go/consensus/tendermint/apps/staking/state/state_test.go (hunk 5 above)
Comment (ptrus) [reply go/consensus/tendermint/apps/staking/state/state_test.go]: fixed (throughout whole file)

File: go/consensus/tendermint/apps/staking/state/state_test.go (hunk 6 above)
Comment (ptrus) [reply go/consensus/tendermint/apps/staking/state/state_test.go]: fixed (throughout whole file)

File: go/consensus/tendermint/apps/staking/state/state_test.go (hunk 7 above)
Comment (ptrus) [reply go/consensus/tendermint/apps/staking/state/state_test.go]: fixed (throughout whole file)

**Comment by ptrus** [comment]
thanks for review!

**Pull request merged**